text/0072-kafka-schema-registry.md

```diff
@@ -0,0 +1,113 @@
+- Start Date: 2023-02-01
+- RFC Type: feature
+- RFC PR: n/a
+- RFC Status: draft
+
+# Summary
+
+This RFC proposes introducing a centralized schema repository for the Kafka topics in use at Sentry. It provides mappings between the Kafka topics we are using to the encoding format and schemas for messages on that topic.
+
+Topic registration and schema validation of messages will be optional for the foreseeable future.
+
+# Motivation
+
+Kafka is increasingly used as the message bus between services at Sentry. Kafka topics and their schemas are usually not internal to any one service, but part of the contract between services.
+
+As Sentry grows, the number of message types and topics in use is exploding. We are marching further towards increasingly distributed ownership of data by product and infrastructure engineering teams. We have already seen an increase in the number of incidents related with invalid data and schema issues at Sentry. It is reasonable to expect this trend to continue as the number of topics, data types, teams and engineers contributing to Sentry grows.
+
+The goals of the centralized schema repository are to:
+
+- provide an explicit, single source of truth about how data in any given topic should look, and the contract that all consumers and producers of a topic must adhere to
+- provide more stability by enforcing backward compatibility of changes via automation
+- explicitly enumerate ownership of schemas
+- make it easier for consumers to identify and reject bad messages without pausing the whole pipeline
+- provide examples of messages for each schema type, which would help with writing consumer tests
+
+# Supporting Data
+
+Over the last quarter, we have seen many incidents related to schema disagreements at Sentry. Examples include the following: (note: inc numbers are internal to Sentry)
+
+- Post process forwarder (INC-218)
+- Snuba’s transactions consumer (INC-210)
+- Super big consumers (INC-220)
+- Replays consumer (INC-250, INC-281)
+
+Many of these incidents are P1. Since messages are in order, an invalid message often halts consumers and requires manual intervention. This is disruptive to both Sentry engineers and our users (it takes us much longer to recover).
+
+# Options Considered
+
+### **Option A (preferred): Publish a library**
```

Comment:
As far as this RFC goes, only the schemas are in scope.

However, I do think we can also use it for the source of truth for other things like topic creation parameters.
If we wanted to add this in the future I think it would not be a difficult addition. It could be something like:
```
topic: events
schemas:
    - version: 1
    type: json
    resource: events.schema.json
topic_creation_config:
    - message.timestamp.type: LogAppendTime
```

Unfortunately I see no easy way to handle any kind of override values here as these are values that can be changed per installation.

Reply:
ok.